contracts/simple_dex/lib.rs

```diff
@@ -0,0 +1,660 @@
+        pub lono: AccountId,
+    }
+
+    impl AccessControlled for SimpleDex {
+        type ContractError = DexError;
+    }
+
+    impl SimpleDex {
+        #[ink(constructor)]
+        pub fn new(ubik: AccountId, cyberiad: AccountId, lono: AccountId) -> Self {
+            let caller = Self::env().caller();
+            let code_hash = Self::env()
+                .own_code_hash()
+                .expect("Called new on a contract with no code hash");
+            let required_role = Role::Initializer(code_hash);
+            let access_control = AccountId::from(ACCESS_CONTROL_PUBKEY);
+
+            let role_check = <Self as AccessControlled>::check_role(
+                access_control,
+                caller,
+                required_role,
+                Self::cross_contract_call_error_handler,
+                Self::access_control_error_handler,
+            );
+
+            match role_check {
+                Ok(_) => {
+                    initialize_contract(|contract| Self::new_init(contract, ubik, cyberiad, lono))
+                }
+                Err(why) => panic!("Could not initialize the contract {:?}", why),
+            }
+        }
+
+        /// Swaps the transferred amount of native token to one of the pools PSP22 tokens
+        #[ink(message, payable)]
+        pub fn native_to_token(&mut self, token_out: AccountId) -> Result<(), DexError> {
+            let amount_token_in = self.env().transferred_value();
+            // we use the balance of the native token just before the exchange tx
+            let balance_token_in = self
+                .env()
+                .balance()
+                .checked_sub(amount_token_in)
+                .ok_or(DexError::Arithmethic)?;
+
+            let swap_fee = self.swap_fee;
+            let caller = self.env().caller();
+            let this = self.env().account_id();
+
+            let balance_token_out = self.balance_of(token_out, this)?;
+
+            let amount_token_out = Self::out_given_in(
+                amount_token_in,
+                balance_token_in,
+                balance_token_out,
+                swap_fee,
+            )?;
+
+            // transfer token_out from contract to user
+            self.transfer_tx(token_out, caller, amount_token_out)??;
+
+            // emit event
+            Self::emit_event(
+                self.env(),
+                Event::NativeToToken(NativeToToken {
+                    caller,
+                    token_out,
+                    amount_in: amount_token_in,
+                    amount_out: amount_token_out,
+                }),
+            );
+
+            Ok(())
+        }
+
+        /// Swap the specified amount of one of the pools PSP22 tokens to the native token
+        /// Calling account needs to give allowance to the DEX contract to spend amount_token_in of token_in on it's behalf
+        /// before executing this tx.
+        #[ink(message)]
+        pub fn token_to_native(
```

Comment:
I can feel your pain, needing to have tons of "ifology" because of the `native` case :P

The standard way to avoid it, is to say: OK let's just have a pool of `4` PSP22 tokens, we don't even care which is which. And we go from native to `PSP22` outside of this contract by "wrapping" `AZERO` into `wAZERO` being a simple PSP22 contract, which allows minting `wAZERO` by depositing `AZERO` and getting back `AZERO` by depositing `wAZERO`. 

It adds some mess to the frontend then, and sometimes requires the user to make a manual wrap/unwrap. But I think it's still state of the art.  

Wrapped AZERO will be probably used all over different contracts anyway. So we could probably deploy the official version, I guess. 

Treat it as a discussion starter, not as a request for changes. The current version is 100% better from UX perspective. Just more complicated when it comes to writing the contract/s, making sure they are correct.